### PR TITLE
feat: handle adyen refusals

### PR DESCRIPTION
**Pull request by Nick-Zag**

### CODE REVIEW

File: packages/demand-bloc-payment/src/providers/AdyenProvider.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import AdyenCheckout from '@adyen/adyen-web'
 import CardElement from '@adyen/adyen-web/dist/types/components/Card'
+import { PaymentAction } from '@adyen/adyen-web/dist/types/types'
```
Comment (aliaksandr-bordak) [inline packages/demand-bloc-payment/src/providers/AdyenProvider.ts:3]: Would shorter version `import { PaymentAction } from '@adyen/adyen-web/dist/types'` do the work?

File: packages/demand-bloc-payment/src/providers/adyenErrors.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  '29': 'Withdrawal count exceeded',
+  '31': 'Issuer Suspected Fraud',
+  '32': 'AVS Declined',
+  '33': 'Card requires online pin',
+  '34': 'No checking account available on Card',
+  '35': 'No savings account available on Card',
+  '36': 'Mobile pin required',
+  '37': 'Contactless fallback',
+  '38': 'Authentication required',
+  '39': 'RReq not received from DS',
```
Comment (aliaksandr-bordak) [inline packages/demand-bloc-payment/src/providers/adyenErrors.ts:41]: Is `RReq` a correct writing? Shouldn't it be `Req`?

File: packages/demand-bloc-payment/src/providers/adyenErrors.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+export const handleRefusalResponse = (result: HttpResponseOk<PaymentAuthResponse>) => {
+  if (result.ok && refusalResultCodes.includes(result.body.payload.resultCode as string)) {
+    const refusalReasonCode: string = result.body.payload.refusalReasonCode as string
```
Comment (aliaksandr-bordak) [inline packages/demand-bloc-payment/src/providers/adyenErrors.ts:46]: At least one of the `string` types is redundant here.
`const refusalReasonCode = result.body.payload.refusalReasonCode as string`
or
`const refusalReasonCode = result.body.payload.refusalReasonCode`

File: packages/demand-bloc-payment/src/providers/adyenErrors.ts (hunk 2 above)
Comment (Nick-Zag) [reply packages/demand-bloc-payment/src/providers/adyenErrors.ts:41]: image

**Review by aliaksandr-bordak — APPROVED** [review]

File: packages/demand-bloc-payment/src/providers/AdyenProvider.ts (hunk 1 above)
Comment (Nick-Zag) [reply packages/demand-bloc-payment/src/providers/AdyenProvider.ts:3]: Unfortunately, no

File: packages/demand-bloc-payment/src/providers/adyenErrors.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import { HttpResponseOk, PaymentAuthResponse } from '@karhoo/demand-api'
+
+export const refusalResultCodes = ['Refused', 'Error', 'Cancelled']
+
+export const refusalReasons: Record<string, string> = {
```
Comment (juliamaksimchik) [inline packages/demand-bloc-payment/src/providers/adyenErrors.ts:5]: why do we need to duplicate these errors? Can we import this from `@adyen-web` package ? What if they add a new error, will we need to add it here as well?

**Review by andrei-oksentyk — APPROVED** [review]

File: packages/demand-bloc-payment/src/providers/adyenErrors.ts (hunk 4 above)
Comment (aliaksandr-bordak) [reply packages/demand-bloc-payment/src/providers/adyenErrors.ts:5]: As far as I understand, adyen npm package doesn't contain API errors definitions.

Comment (Nick-Zag) [reply packages/demand-bloc-payment/src/providers/adyenErrors.ts:5]: There are no errors description in @adyen-web package, so we need to keep them in our package

**Pull request merged**